eth/watchers/pricefeedwatcher.go

```diff
@@ -0,0 +1,176 @@
+		ctx:            ctx,
+		baseRetryDelay: priceUpdateBaseRetryDelay,
+		priceFeed:      priceFeed,
+		currencyBase:   currencyFrom,
+		currencyQuote:  currencyTo,
+		priceUpdated:   make(chan eth.PriceData, 1),
+	}
+
+	err = w.updatePrice()
+	if err != nil {
+		return nil, fmt.Errorf("failed to update price: %w", err)
+	}
+
+	go func() {
+		ctx, cancel := context.WithCancel(w.ctx)
+		defer cancel()
+		ticker := newTruncatedTicker(ctx, priceUpdatePeriod)
+		w.watch(ctx, ticker)
+	}()
+
+	return w, nil
+}
+
+// Currencies returns the base and quote currencies of the price feed.
+// i.e. base = CurrentPrice() * quote
+func (w *PriceFeedWatcher) Currencies() (base string, quote string) {
+	return w.currencyBase, w.currencyQuote
+}
+
+// Current returns the latest fetched price data.
+func (w *PriceFeedWatcher) Current() eth.PriceData {
+	return w.current
+}
+
+// PriceUpdated returns a channel that will receive the updated price data as
+// soon as it changes.
+func (w *PriceFeedWatcher) PriceUpdated() <-chan eth.PriceData {
+	return w.priceUpdated
+}
+
+func (w *PriceFeedWatcher) updatePrice() error {
+	newPrice, err := w.priceFeed.FetchPriceData()
+	if err != nil {
+		return fmt.Errorf("failed to fetch price data: %w", err)
+	}
+
+	if newPrice.UpdatedAt.After(w.current.UpdatedAt) {
+		w.current = newPrice
+		select {
+		case w.priceUpdated <- newPrice:
+		default:
+		}
+	}
+
+	return nil
+}
+
+func (w *PriceFeedWatcher) watch(ctx context.Context, ticker <-chan time.Time) {
+	for {
+		select {
+		case <-w.ctx.Done():
+			return
+		case <-ticker:
+			attempt, retryDelay := 1, w.baseRetryDelay
+			for {
+				err := w.updatePrice()
+				if err == nil {
+					break
+				} else if attempt >= priceUpdateMaxRetries {
+					clog.Errorf(ctx, "Failed to fetch updated price from PriceFeed attempts=%d err=%q", attempt, err)
+					break
+				}
+
+				clog.Warningf(ctx, "Failed to fetch updated price from PriceFeed, retrying after retryDelay=%d attempt=%d err=%q", retryDelay, attempt, err)
+				time.Sleep(retryDelay)
+				attempt, retryDelay = attempt+1, retryDelay*2
+			}
+		}
+	}
+}
+
+// parseCurrencies parses the base and quote currencies from a price feed based
+// on Chainlink PriceFeed description pattern "FROM / TO".
+func parseCurrencies(description string) (currencyBase string, currencyQuote string, err error) {
+	currencies := strings.Split(description, "/")
+	if len(currencies) != 2 {
+		return "", "", fmt.Errorf("aggregator description must be in the format 'FROM / TO' but got: %s", description)
+	}
+
+	currencyBase = strings.TrimSpace(currencies[0])
+	currencyQuote = strings.TrimSpace(currencies[1])
+	return
+}
+
+// newTruncatedTicker creates a ticker that ticks at the next time that is a
+// multiple of d, starting from the current time.
+func newTruncatedTicker(ctx context.Context, d time.Duration) <-chan time.Time {
+	return newTruncatedTickerMockable(ctx, d, time.Now, time.After)
+}
+
+// Test helper so we can mock the time functions in tests.
```

Comment:
I'd try to avoid making the logic (so much more) complex just because of unit testing.

Reply:
Yeah this was quite a pain to test TBH. Do you think I should just drop those tests?

Reply:
TBH I'd probably drop the tests.

Reply:
dropped truncated ticker tests! the watch ones seemed worth to keep, mocking the ticker didnt seem that weird